### PR TITLE
ESM script base class

### PR DESCRIPTION
**Pull request by marklundin**
This PR proposes a new `Script `base class for ESM Scripts. Attributes are declared via jsdoc tags as opposed to  `Script.attributes.add()`. `ScriptType` class persists for legacy support. The new `Script` base class:

- does **not** use the `ScriptAttributes` class
- does **not** override class members with getters/setters. Fixes https://github.com/playcanvas/editor/issues/1148
- is compatible with [JSDoc style attributes](https://github.com/playcanvas/editor/issues/1148)
- is assigned any attributes based upon the json schema in it's corresponding asset

## API Changes

 - New ESM `Script` Class
A new minimal base. Attributes are declared via jsdoc tags. Runtime attributes via `Script.attributes.add()` is removed. Fixes https://github.com/playcanvas/engine/issues/6316

- `assignRawToValue` - New method that takes a script instance, a schema and value and assigns a new value on the script instance`
- `ScriptRegistry.addSchema`/`getSchema` - Stores and fetches an attributes schema from it's associated asset `<scriptName, schema>`

## Files renaming
For consistency, the following files have been renamed
`script.js` -> `script-create.js` - Contains legacy `pc.createScript()` and `pc.registerScript()`
`script.js` contains new `Script` base class 
`script-type.js` contains legacy `ScriptType` class

### Tests
`npm run test:karma`

### CODE REVIEW

**Comment by Maksims** [comment]
What would be the impact to users in terms of migrations or API changes?

**Comment by marklundin** [comment]
> What would be the impact to users in terms of migrations or API changes?

`pc.createScript()` will still work and return a `ScriptType` which should function in the same way. ESM users will use `Script` as a base class without any ScriptAttribute functionality (events, `rawToValue` deep copying)

File: src/framework/script/script-create.js
Comment (marklundin) [reply src/framework/script/script-create.js:74]: This is legacy code, and it's not clear why ScriptType are initialized this way. This was before ES classes and `super` syntax, so it's likely this is trying to polyfill some functionality. I don't really want to change how this works or modify anything related to ScriptType as we will eventually deprecate this

File: src/framework/script/script-create.js
Comment (marklundin) [reply src/framework/script/script-create.js:60]: `pc.createScript` is existing API and I'm not keen on changing as we'd need to migrate all existing projects.

File: src/framework/script/script-create.js
Comment (LeXXik) [reply src/framework/script/script-create.js:24]: Speaking of which - the Editor UI should not allow creating a script with a reserved name, e.g. "script".

Comment (marklundin) [reply src/framework/script/script-create.js:24]: This is not part of this PR, it's just a renamed file https://github.com/playcanvas/engine/blob/main/src/framework/script/script.js#L11-L23

Comment (LeXXik) [reply src/framework/script/script-create.js:24]: Yeah, I should create a separate issue in Editor repo.

File: src/framework/components/script/component.js
Comment (kpal81xd) [inline src/framework/components/script/component.js]:
```suggestion
```

File: src/framework/components/script/component.js
Comment (kpal81xd) [reply src/framework/components/script/component.js:349]: +1

File: src/framework/components/script/component.js
Comment (kpal81xd) [inline src/framework/components/script/component.js:415]:
```suggestion
            if (!data) {
                return;
            }
```

File: src/framework/components/script/component.js
Comment (kpal81xd) [inline src/framework/components/script/component.js:419]:
```suggestion
            if (!schema) {
                Debug.warnOnce(`No schema exists for the script '${name}'. A schema must exist for data to be instantiated on the script.`);
            }
```

File: src/framework/components/script/component.js
Comment (mvaligursky) [inline src/framework/components/script/component.js:410]: there is already a function with this name on top of the file

File: src/framework/components/script/component.js
Comment (mvaligursky) [inline src/framework/components/script/component.js:26]: should it be `initial` instead of `initialized`?

File: src/framework/components/script/component.js
Comment (kpal81xd) [inline src/framework/components/script/component.js:399]: This can be folded down into one function

File: src/framework/components/script/component.js
Comment (marklundin) [reply src/framework/components/script/component.js]: Yep just caught this. Fixed

File: src/framework/components/script/component.js
Comment (mvaligursky) [inline src/framework/components/script/component.js:205]: not related to this PR directly, but the function says it's read only .. so perhaps we should not have a setter at all? Only a public getter.
And this setter could be changed to something like `setScripts` which is `@ignore`

**Review by kpal81xd — DISMISSED** [review]

File: src/framework/components/script/component.js
Comment (mvaligursky) [inline src/framework/components/script/component.js:411]: it feels like this should be a function as this is in the code base multiple times.

File: src/framework/script/script-attributes.js
Comment (mvaligursky) [inline src/framework/script/script-attributes.js:155]: are `step`, `min` and `max` members used by the engine?

File: src/framework/script/script-registry.js
Comment (mvaligursky) [inline src/framework/script/script-registry.js:45]:
```suggestion
        this.scriptSchemas ??= new Map();
```

File: src/framework/script/script-registry.js
Comment (mvaligursky) [inline src/framework/script/script-registry.js:54]: is this part `schema[script]` needed here? isn't the line above already doing lookup by the `script`?
this whole function could perhaps be just
```
return this.scriptSchemas?.get(script);
```

**Review by mvaligursky — APPROVED** [review]
Looks good to me, added few small comments.
But I'd suggest some additional approvals before merging.
Please make sure no unwanted functions are exposed in the public API here.

File: src/framework/components/script/component.js
Comment (marklundin) [reply src/framework/components/script/component.js:205]: can look at this in followup PR

File: src/framework/script/script-attributes.js
Comment (marklundin) [reply src/framework/script/script-attributes.js:155]: nope. removing

File: src/framework/script/script-registry.js
Comment (marklundin) [reply src/framework/script/script-registry.js:54]: yep, valid

File: src/framework/components/script/component.js
Comment (marklundin) [reply src/framework/components/script/component.js]: confusingly 'enable' can actually fire when this condition is false, so once wont work. This code can be clearer though, so I will refactor

**Comment by Maksims** [comment]
Can we provide a few examples of old-vs-new ways for most common use cases?
As I understand, this is fully compatible with the current ScriptComponent system and can mix with legacy scripts within the component, so devs can gradually migrate current scripts to ESM scripts?

Also, as there are expected few other systems and improvements for ESM (engine and editor), would it be beneficial to mark this API in a way that devs know it is still in the works? So the community can try and battle test the new ESM system fully, before it is a recommended way for everyone?

File: src/framework/components/script/component.js
Comment (marklundin) [reply src/framework/components/script/component.js]: done

File: src/framework/script/script-create.js
Comment (kpal81xd) [inline src/framework/script/script-create.js:69]: Im prettys sure I wrote some code to update this at some point but probably work adding an a later PR

**Review by kpal81xd — APPROVED** [review]

File: src/framework/script/script-attributes.js
Comment (willeastcott) [inline src/framework/script/script-attributes.js:170]:
```suggestion
        return value.map((item, index) => rawToValue(app, schema, item, current ? current[index] : null));
```

File: src/framework/script/script-attributes.js
Comment (willeastcott) [inline src/framework/script/script-attributes.js:177]: Limit JSDoc blocks to 100 characters in width unless there's a good reason not to. And end sentences with a full stop.

File: src/framework/script/script-registry.js
Comment (willeastcott) [inline src/framework/script/script-registry.js:24]: The description goes above the `@type` tag. And our convention is to leave a blank link between the description and the first tag line.

File: src/framework/script/script-registry.js
Comment (willeastcott) [inline src/framework/script/script-registry.js:46]: Leave a blank line after the description.

File: src/framework/script/script-registry.js
Comment (willeastcott) [inline src/framework/script/script-registry.js:143]: Since this isn't performance critical code, I'd just do:

```
                for (const script of component.scripts) {
                    component.initializeAttributes(script);
                }
```

File: src/framework/script/script-type.js
Comment (willeastcott) [inline src/framework/script/script-type.js:49]: Is this intended to be public API in the API ref manual?

File: src/framework/script/script.js
Comment (willeastcott) [inline src/framework/script/script.js:11]: This is a bit confusing. A script can listen for the `enable` event. Which implies a script it initialized and _then_ enabled.

File: src/framework/script/script.js
Comment (willeastcott) [inline src/framework/script/script.js:46]: This is scripts v2 format. Should the example code be ESM script format?

Comment (willeastcott) [reply src/framework/script/script.js:46]: Same for other events below.

File: src/framework/script/script.js
Comment (willeastcott) [inline src/framework/script/script.js:243]: The JSDoc suggests that only `app` and `entity` can appear as properties of `args`. But this suggests `enabled` can be present too.

File: src/framework/script/script-type.js
Comment (marklundin) [reply src/framework/script/script-type.js:49]: Nope, marking as @protected

File: src/framework/script/script.js
Comment (marklundin) [reply src/framework/script/script.js:243]: Added separate typedef to complete the init args

File: src/framework/components/script/component.js
Comment (willeastcott) [inline src/framework/components/script/component.js:28]:
```suggestion
     * A map of script name to initial component data.
     *
     * @type {Map<string, object>}
     * @private
```

File: src/framework/script/script-attributes.js
Comment (willeastcott) [inline src/framework/script/script-attributes.js:156]:
```suggestion
 * Takes an attribute schema, a value and current value, and return a new value.
```

File: src/framework/script/script-attributes.js
Comment (willeastcott) [inline src/framework/script/script-attributes.js:164]: Just wanted to check...have you taken a look at the generated API reference to see whether any JSDoc blocks are unexpectedly being made public? Hard to tell if, say, this one is...

File: src/framework/script/script-type.js
Comment (willeastcott) [inline src/framework/script/script-type.js:39]:
```suggestion
     * @param {*} args - initialization arguments
     * @protected
```

File: src/framework/script/script.js
Comment (willeastcott) [inline src/framework/script/script.js:7]: * The `Script` class is the fundamental base class for all scripts within PlayCanvas. It provides
 * the minimal interface required for a script to be compatible with both the Engine and the
 * Editor.

Comment (willeastcott) [inline src/framework/script/script.js:7]:
```suggestion
 * The `Script` class is the fundamental base class for all scripts within PlayCanvas. It provides
 * the minimal interface required for a script to be compatible with both the Engine and the
 * Editor.
```

File: src/framework/script/script.js
Comment (willeastcott) [inline src/framework/script/script.js:9]:
```suggestion
 * At its core, a script is simply a collection of methods that are called at various points in the Engine's lifecycle. These methods are:
```

File: src/framework/script/script.js
Comment (willeastcott) [inline src/framework/script/script.js:129]: My suspicion is that `@link` just work for internal symbols. It won't link to, say, MDN. But feel free to test.

File: src/framework/script/script.js
Comment (willeastcott) [inline src/framework/script/script.js:327]:
```suggestion
     * @description Called when a Script that already exists in the registry gets redefined. If the
     * new Script has a `swap` method, then it will be executed to perform hot-reload at runtime.
```

**Review by willeastcott — APPROVED** [review]
Approving with a final round of comments.

File: src/framework/script/script-attributes.js
Comment (marklundin) [reply src/framework/script/script-attributes.js:164]: Yep, this no longer needs to be public. Removing the export

**Pull request merged**
